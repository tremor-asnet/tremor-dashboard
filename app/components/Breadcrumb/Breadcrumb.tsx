// Libs
import { memo, useMemo } from "react";
import Link from "next/link";
import { usePathname } from "next/navigation";
import { Title } from "@tremor/react";
import { MdHome } from "react-icons/md";

// Constants
import { ROUTES } from "../../constants";

// Types
import { IBreadcrumb } from "@/types";

// Helpers
import { isBrowser } from "@/helpers";

/**
 * Primary UI component for BreadCrumb component
 */

export interface BreadcrumbProps {
  isScrolled?: boolean;
}

const Breadcrumb = ({ isScrolled = false }: BreadcrumbProps): JSX.Element => {
  const pathname = usePathname();
  const isMobile = isBrowser && window.innerWidth <= 768;
  const isProjectPage = pathname === ROUTES.PROJECTS;
  const isStickyHeader = !isScrolled && isProjectPage;

  const links: IBreadcrumb[] = useMemo(() => {
    switch (true) {
      case pathname?.includes("/dashboards/"):
        return [{ name: "dashboards", url: ROUTES.ANALYTICS }];

      case pathname?.includes("/pages/profile/"):
        return [
          { name: "pages", url: ROUTES.PROFILE },
          { name: "profile", url: ROUTES.PROFILE },
        ];

      case pathname?.includes("/pages/account/"):
        return [
          { name: "pages", url: ROUTES.PROFILE },
          { name: "account", url: ROUTES.PROFILE },
        ];

      default:
        return [{ name: "dashboards", url: ROUTES.ANALYTICS }];
    }
  }, [pathname]);

  const pageName = useMemo(() => {
    switch (pathname) {
      case ROUTES.ANALYTICS:
        return "analytics";
      case ROUTES.SALES:
        return "sales";
      case ROUTES.PROFILE:
        return "my profile";
      case ROUTES.PROJECTS:
        return "all projects";
      case ROUTES.SETTING:
        return "settings";
      case ROUTES.PRODUCT_LIST:
        return "product-list";
      default:
        return "analytics";
    }
  }, [pathname]);

  const renderLinks = (): JSX.Element[] => {
    return links.map(link => (
      <li
        key={link.name}
        className={`flex items-center ${
          isStickyHeader ? "text-white opacity-[0.8]" : "text-inherit"
        }`}>
        <div className="bc-link">
          <Link
            className={`text-sm capitalize tracking-[0.02857em] leading-[0] ${
              isStickyHeader
                ? "text-white opacity-[0.8]"
                : "text-primary dark:text-dark-primary opacity-50"
            }`}
            href={link.url}>
            {link.name}
          </Link>
        </div>
        <div
          aria-hidden="true"
          className={`text-sm mx-2 ${
            isStickyHeader ? "text-white" : "text-[#6c757d]"
          }`}>
          &#47;
        </div>
      </li>
    ));
  };

  return (
    <nav className={`${isProjectPage ? "pl-3 z-20" : isMobile ? "pl-3" : ""}`}>
      <ol className="flex flex-wrap items-center text-gray-400">
        <li>
          <Link href={ROUTES.HOME} className="flex">
            <MdHome
              className={`bg-inherit border-0 py-0 hover:bg-transparent text-primary dark:text-dark-primary tracking-[0.01071em] leading-[0] ${
                isStickyHeader
                  ? "text-white opacity-[0.8]"
                  : "text-inherit opacity-50"
              }`}
            />
          </Link>
        </li>
        <li
          aria-hidden="true"
          className={`text-sm mx-2 ${
            isStickyHeader ? "text-white" : "text-[#6c757d]"
          }`}>
          &#47;
        </li>

        {renderLinks()}

        <li
          className={`text-sm text-tremor-content-title dark:text-dark-tremor-content-title capitalize tracking-[0.02857em] leading-[0] ${
            isStickyHeader ? "text-white" : "text-inherit"
          }`}>
          {pageName}
        </li>
      </ol>
      <Title
        className={`text-tremor-content-title dark:text-dark-tremor-content-title font-bold capitalize tracking-[0.0075em] ${
          isStickyHeader ? "text-white" : "text-primary"
        }`}>
        {pageName}
      </Title>
    </nav>
  );
};

export default memo(Breadcrumb);
