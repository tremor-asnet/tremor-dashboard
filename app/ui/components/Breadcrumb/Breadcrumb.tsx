// Libs
import { useParams } from "next/navigation";
import { memo } from "react";
import Link from "next/link";

// Components
import { Title } from "@tremor/react";
import { MdHome } from "react-icons/md";

// Constants
import { ROUTES } from "@/constants";

// Helpers
import { getCrumbName, handleMatchPath, isBrowser } from "@/helpers";

export interface BreadcrumbProps {
  isScrolled?: boolean;
  pathname: string;
}

const Breadcrumb = ({
  isScrolled = false,
  pathname,
}: BreadcrumbProps): JSX.Element => {
  const params = useParams();
  const isMobile = isBrowser && window.innerWidth <= 768;
  const isProjectPage = pathname === ROUTES.PROJECTS;
  const isStickyHeader = !isScrolled && isProjectPage;
  const newPath = pathname?.split("/").filter(path => path);

  const renderTitle = (path?: string) => {
    if (path) {
      const matchPath = handleMatchPath(path);

      return matchPath;
    }

    return newPath?.pop()?.replace("-", " ");
  };

  const renderCrumb = () => {
    return (
      <>
        {newPath?.map((link, index) => {
          const href = `/${newPath.slice(0, index + 1).join("/")}`;
          const crumbItemTextWhite =
            isProjectPage && isStickyHeader
              ? "text-white dark:text-white"
              : "text-primary";
          const activeCrumb =
            pathname === href
              ? "text-tremor-content-title dark:text-dark-tremor-content-title"
              : "text-primary dark:text-dark-primary opacity-50";

          return (
            <li
              key={href}
              className={`flex gap-1 sm:gap-2 bc-link text-sm capitalize tracking-[0.02857em] ${crumbItemTextWhite} ${activeCrumb}`}>
              <Link href={href}>
                {getCrumbName({
                  name: link,
                  path: href,
                  params: params?.id,
                })}
              </Link>
              {newPath.length !== index + 1 && (
                <span className={`${isProjectPage && "text-white"}`}>
                  &#47;
                </span>
              )}
            </li>
          );
        })}
      </>
    );
  };

  return (
    <nav className={`${isProjectPage ? "pl-3 z-20" : isMobile ? "pl-3" : ""}`}>
      <ol
        className={`flex flex-wrap gap-1 sm:gap-2 items-center ${
          isProjectPage && isStickyHeader
            ? "text-lighter"
            : "text-gray-400 dark:text-white"
        }`}>
        <li className="text-sm flex items-center gap-1 sm:gap-2 capitalize dark:text-dark-primary tracking-[0.02857em]">
          <Link href={ROUTES.HOME} aria-label="Home icon">
            <MdHome className="text-lg" />
          </Link>
          &#47;
        </li>
        {renderCrumb()}
      </ol>
      <Title
        className={`text-tremor-content-title dark:text-dark-tremor-content-title font-bold capitalize tracking-[0.0075em] mt-1 ${
          isStickyHeader ? "text-white" : "text-primary"
        }`}>
        {renderTitle(pathname)}
      </Title>
    </nav>
  );
};

export default memo(Breadcrumb);
