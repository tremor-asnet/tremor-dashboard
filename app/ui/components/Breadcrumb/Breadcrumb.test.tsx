import { render } from "@testing-library/react";

// Components
import Breadcrumb from "./Breadcrumb";

// Mocking usePathname hook
jest.mock("next/navigation", () => ({
  usePathname: jest
    .fn()
    .mockReturnValueOnce("/all-projects")
    .mockReturnValueOnce("/product-list")
    .mockReturnValue("/product-list/87120/edit-product"),
  useParams: jest.fn().mockReturnValue({}),
}));

describe("Testing breadcrumb component", () => {
  it("Should match snapshot", () => {
    const { container } = render(
      <Breadcrumb isScrolled={true} pathname={""} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("Should render correctly with display breadcrumb with route projects", () => {
    const { container } = render(
      <Breadcrumb isScrolled={false} pathname={"/all-projects/8024"} />,
    );
    const link = container.getElementsByClassName("bc-link")[0];
    expect(link.textContent).toEqual("all projects/");
  });

  it("Should render correctly with display breadcrumb text", () => {
    const { container } = render(
      <Breadcrumb isScrolled={false} pathname={"/product-list"} />,
    );
    const link = container.getElementsByClassName("bc-link")[0];
    expect(link.textContent).toEqual("product list");
  });

  it("Should render correctly with display breadcrumb more details", () => {
    const { container } = render(
      <Breadcrumb
        isScrolled={false}
        pathname={"/product-list/87120/edit-product"}
      />,
    );
    const link = container.getElementsByClassName("bc-link")[2];
    expect(link.textContent).toEqual("edit product");
  });
});
