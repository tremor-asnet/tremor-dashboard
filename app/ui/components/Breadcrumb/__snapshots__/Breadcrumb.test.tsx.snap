// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing breadcrumb component Should match snapshot 1`] = `
<div>
  <nav
    class=""
  >
    <ol
      class="flex flex-wrap gap-1 sm:gap-2 items-center text-gray-400 dark:text-white"
    >
      <li
        class="text-sm flex items-center gap-1 sm:gap-2 capitalize dark:text-dark-primary tracking-[0.02857em]"
      >
        <a
          aria-label="Home icon"
          href="/analytics"
        >
          <svg
            class="text-lg"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </a>
        /
      </li>
      <li
        class="flex gap-1 sm:gap-2 bc-link text-sm capitalize tracking-[0.02857em] text-primary text-tremor-content-title dark:text-dark-tremor-content-title"
      >
        <a
          href="/analytics"
        >
          analytics
        </a>
      </li>
    </ol>
    <p
      class="text-tremor-title dark:text-dark-tremor-content-title font-bold capitalize tracking-[0.0075em] mt-1 text-primary"
    />
  </nav>
</div>
`;
